x-pack/plugin/esql/src/internalClusterTest/java/org/elasticsearch/xpack/esql/action/LookupJoinTypesIT.java

```diff
@@ -0,0 +1,463 @@
+                    if (existingIndex(existing, mainType, lookupType)) {
+                        // Skip existing configurations
+                        continue;
+                    }
+                    configs.addFails(mainType, lookupType);
+                }
+            }
+        }
+
+        // TODO: Add tests for more types, eg. unsigned_long, version, spatial types, date/temporal types.
+
+        // Make sure we have never added two configurations with the same index name
+        Set<String> knownTypes = new HashSet<>();
+        for (TestConfigs configs : testConfigurations.values()) {
+            for (TestConfig config : configs.configs.values()) {
+                if (knownTypes.contains(config.indexName())) {
+                    throw new IllegalArgumentException("Duplicate index name: " + config.indexName());
+                }
+                knownTypes.add(config.indexName());
+            }
+        }
+    }
+
+    private static boolean existingIndex(Collection<TestConfigs> existing, DataType mainType, DataType lookupType) {
+        String indexName = "index_" + mainType.esType() + "_" + lookupType.esType();
+        return existing.stream().anyMatch(c -> c.exists(indexName));
+    }
+
+    public void testLookupJoinStrings() {
+        testLookupJoinTypes("strings");
+    }
+
+    public void testLookupJoinIntegers() {
+        testLookupJoinTypes("integers");
+    }
+
+    public void testLookupJoinFloats() {
+        testLookupJoinTypes("floats");
+    }
+
+    public void testLookupJoinMixedNumerical() {
+        testLookupJoinTypes("mixed-numerical");
+    }
+
+    public void testLookupJoinSame() {
+        testLookupJoinTypes("same");
+    }
+
+    public void testLookupJoinOthers() {
+        testLookupJoinTypes("others");
+    }
+
+    private void testLookupJoinTypes(String group) {
+        initIndexes(group);
+        initData(group);
+        for (TestConfig config : testConfigurations.get(group).configs.values()) {
+            String query = String.format(
+                Locale.ROOT,
+                "FROM index | LOOKUP JOIN %s ON %s | KEEP other",
+                config.indexName(),
+                config.fieldName()
+            );
+            config.validateMainIndex();
+            config.validateLookupIndex();
+            config.testQuery(query);
+        }
+    }
+
+    private void initIndexes(String group) {
+        Collection<TestConfig> configs = testConfigurations.get(group).configs.values();
+        // The main index will have many fields, one of each type to use in later type specific joins
+        String mainFields = "{\n  \"properties\" : {\n"
+            + configs.stream().map(TestConfig::mainPropertySpec).distinct().collect(Collectors.joining(",\n    "))
+            + "  }\n}\n";
+        assertAcked(prepareCreate("index").setMapping(mainFields));
+
+        Settings.Builder settings = Settings.builder()
+            .put("index.number_of_shards", 1)
```

Comment:
This is theoretically redundant; we used to have to set this manually, but by now I think just setting the index mode to lookup should suffice. I think @idegtiarenko double checked this the other day.

Reply:
In the off-chance that we change that sometime in the future, I think perhaps this test should keep this, as the test should focus on type-compatibilities, and not be influenced by any possible changes in lookup shard counts.